src/daft-functions/src/python/mod.rs

```diff
@@ -64,30 +67,27 @@ impl PyScalarFunction {
             }
         }
 
-        let expr: ExprRef = ScalarFunction {
-            udf: self.inner.clone(),
-            inputs: FunctionArgs::try_new(inputs)?,
-        }
-        .into();
+        // Python only uses dynamic function, so we pass an empty schema.
+        // Once ExprRef uses the ScalarFunctionFactory, then we can do function
+        // resolution in the planner. For now, we'll necessarily get the single dynamic function
```

Comment:
If we do decide to move forward with these function factories, is there any way we can make it opt-in? My main concern is that this adds another layer to making expressions. I'm not convinced that the currying functionality is worth the additional complexity.

Reply:
That's a fair concern. I've set it up so that the existing and documented `module.register_fn` will create the single/dynamic factory for you. So no additional complexity for the users, and you don't actually have to do anything which is part of why this PR is so small. This merely introduces the indirection between name and type resolution which subtly existed before, and this sets us up for type resolution later.